jupyter-extension/jupyterlab_pachyderm/pps_client.py

```diff
@@ -41,72 +39,142 @@ async def create(self, path: str, config: dict):
-            # Remove requirements field if none are specified.
-            del config['notebook']['requirements']
-
-        with NamedTemporaryFile() as temp_config:
-            Path(temp_config.name).write_text(json.dumps(config))
-            command = [sys.executable, "-m", "sameproject.main", "run"]
-            command.extend(["--same-file", temp_config.name])
-            command.extend(["--target", "pachyderm"])
-            command.extend(["--input", input_spec])
-            call = subprocess.run(command, capture_output=True)
-            if call.returncode:
-                raise HTTPError(
-                    code=500,
-                    log_message=call.stderr.decode(),
-                    reason="Internal Error: SAME could not create pipeline"
-                )
+        config = PpsConfig.from_json(config)
+        path = path.lstrip("/")
+        if not os.path.exists(path):
+            raise HTTPError(reason=f"notebook does not exist: {path}")
+        if config.requirements and not os.path.exists(config.requirements):
+            raise HTTPError(reason="requirements file does not exist")
+
+        script, _resources = self.nbconvert.from_filename(path)
+
+        client = python_pachyderm.Client()  # TODO: Auth?
+        companion_repo = f"{config.pipeline_name}__context"
+        if companion_repo not in (item.repo.name for item in client.list_repo()):
+            client.create_repo(
+                repo_name=companion_repo,
+                description=f"files for running the {config.pipeline_name} pipeline"
+            )
+
+        companion_branch = upload_environment(client, companion_repo, config, script.encode('utf-8'))
+        pipeline_spec = create_pipeline_spec(config, companion_branch)
+        client.create_pipeline_from_request(
+            python_pachyderm.parse_dict_pipeline_spec(pipeline_spec)
+        )
 
         return json.dumps(
             dict(message=None)  # We can send back console link here.
         )
 
 
-def create_pipeline_spec(
-    pipeline_name: str,
-    image: str,
-    input_spec: dict,
-    script_name: str,
-) -> dict:
-    """Generates the pipelines spec. Currently copied from """
-    companion_repo = f"{pipeline_name}__context"
-    micro_entrypoint = (
-        'print("Greetings from the Pachyderm PPS Extension"); '
-        + 'import sys; '
-        + 'from importlib import import_module; '
-        + 'from pathlib import Path; '
-        + 'from subprocess import run; '
-
-        + 'root_module = sys.argv[1]; '
-        + f'context_dir = Path("/pfs", "{companion_repo}"); '
-        + 'reqs = context_dir / "requirements.txt"; '
-        + 'reqs.exists() and run(["pip", "--disable-pip-version-check", "install", "-r", reqs.as_posix()]); '
-        + 'sys.path.append(context_dir.as_posix()); '
-        + 'script = import_module(root_module); '
-        + 'script.root()'
+@dataclass
+class PpsConfig:
```

Comment:
Not sure of the python convention here w.r.t. separate classes, but should this be in it's own file?

Reply:
It's generally ok to keep related classes in the same file.

Reply:
Future work: We also might want to consider separating concerns here. Generic PPSConfig class which returns an error, and then the caller can decide what to do with it (wrap in http, etc)

Reply:
Oh I actually did exactly this in #8696